### DISCUSSION TITLE
Chakra typography defaults (similar to Tailwind Typography)?

### DISCUSSION THREAD

**Discussion by PrimozRome** [chakra-ui/chakra-ui#5566] [Q&A] [answered]
Does Chakra UI provide something like [Tailwind Typography](https://tailwindcss.com/docs/typography-plugin)? If not, how do you handle, for example, the default styles for HTML that comes from a database (e.g. blog post)?

_Beautiful typographic defaults for HTML you don't control. The official Tailwind CSS Typography plugin provides a set of prose classes you can use to add beautiful typographic defaults to any vanilla HTML you don’t control, like HTML rendered from Markdown, or pulled from a CMS_.

**Comment by nikolovlazar** [comment]
Hi @PrimozRome! There's no single component built-in Chakra UI that does this. If you're using Markdown/MDX you can provide your own components like I have in my website:
- MDX components: https://github.com/nikolovlazar/nikolovlazar.com/blob/main/src/components/mdx-components/index.tsx
- Blog page: https://github.com/nikolovlazar/nikolovlazar.com/blob/main/pages/blog/%5Bslug%5D.tsx#L99

**Comment by PrimozRome** [comment]
@nikolovlazar thank you for your input. We are not using Markdown but reach HTML editor in the Headless CMS... We are getting RAW html from our site...

I checked your solution with MDX, looks good, but it doesn't solve my problem.

**Reply by nikolovlazar** [reply]
There's another hack as well. Wrap your content with a `Box` component and provide the styles in the `sx` prop:
```typescript
<Box
   sx={{
      h1: {
         fontSize: '4xl',
         fontWeight: 'bold',
         ...
      }
      ...
   }}
>
   {content}
</Box>
```

I'll look into building something ready-to-use similar to Tailwind's Typography plugin. I'll try to remember to ping you when I publish it, but keep an eye just in case 😅

**Reply by PrimozRome** [reply]
I will try play with the suggested hack temporary. 

> I'll look into building something ready-to-use similar to Tailwind's Typography plugin. I'll try to remember to ping you when I publish it, but keep an eye just in case 😅

That sounds awesome. I think this will be good addition! Appreciated a lot!

**Reply by PrimozRome** [reply]
@nikolovlazar Hi there. Did you maybe find some time to work on this?

**Reply by nikolovlazar** [reply]
Yes I did, @PrimozRome! It'll be a brand new `Prose` component, and it's almost done. You can check out the [feature/prose-component](https://github.com/chakra-ui/chakra-ui/tree/feature/prose-component) branch to see the progress. I've set up a [Storybook Example](https://github.com/chakra-ui/chakra-ui/blob/feature/prose-component/packages/prose/stories/prose.stories.tsx) that uses the new component.

If you have the `chakra-ui` repo cloned locally, you can do `yarn storybook` to boot up Storybook and check the example. I need to do some more styling, a refactoring, and the new component will be published!

Check out the example and let me know if I'm missing any element styling. If you have any ideas, I'm open to discussion.

Thanks!

**Comment by PrimozRome** [comment]
That is awesome @nikolovlazar. I am trying to run the storybook example, will let you know my feedback when I succeed.

**Reply by PrimozRome** [reply]
Hi again @nikolovlazar 

_If you have the chakra-ui repo cloned locally, you can do yarn storybook to boot up Storybook and check the example. I need to do some more styling, a refactoring, and the new component will be published!_

I tried this but get errors. Bellow is the first error. I am probably doing something wrong as I have never used the Storybook before. I have cloned `chakra-ui` repo and run `yarn storybook`

```
ERROR in ./packages/menu/src/use-menu.ts
Module not found: Error: Can't resolve '@chakra-ui/hooks/use-animation-state' in '/Users/primozrome/Documents/Web/chakra-ui/packages/menu/src'
 @ ./packages/menu/src/use-menu.ts 31:0-73 189:23-40
 @ ./packages/menu/src/index.ts
 @ ./packages/menu/stories/menu.stories.tsx
 @ ./packages sync ^\.(?:(?:^|\/|(?:(?:(?!(?:^|\/)\.).)*?)\/)stories\/(?!\.)(?=.)[^/]*?\.stories\.tsx)$
 @ ./.storybook/generated-stories-entry.js
 @ multi ./node_modules/@storybook/core-client/dist/esm/globals/polyfills.js ./node_modules/@storybook/core-client/dist/esm/globals/globals.js ./.storybook/storybook-init-framework-entry.js ./node_modules/@storybook/addon-a11y/dist/esm/a11yRunner.js-generated-other-entry.js ./node_modules/@storybook/addon-a11y/dist/esm/a11yHighlight.js-generated-other-entry.js ./.storybook/preview.tsx-generated-config-entry.js ./.storybook/generated-stories-entry.js (webpack)-hot-middleware/client.js?reload=true&quiet=false&noInfo=undefined

```

**Reply by nikolovlazar** [reply]
Yeah you need to install the packages first. Check out the [Contributing docs](https://github.com/chakra-ui/chakra-ui/blob/main/CONTRIBUTING.md) to get started.

**Reply by PrimozRome** [reply]
@nikolovlazar I made Storybook work and checked the new `Prose` component. I don't see any formatting, did I do something wrong? See the screenshot from Storybook:

![chakra-ui-prose-component](https://user-images.githubusercontent.com/101252401/157441680-5e941508-f641-455e-a8b9-1a4e153d7171.jpg)

**Reply by nikolovlazar** [reply]
@PrimozRome did you run `yarn build` before you ran `yarn storybook`?

**Comment by PrimozRome** [comment]
@nikolovlazar yes sure I did, but it always stopped at this point:

```
...
🎁 info @chakra-ui/styled-system created packages/styled-system in 19s
🎁 info @chakra-ui/styled-system waiting for changes...
```

Tried few times but had to manually stop the build script because it would get stuck at above line for hours.

**Reply by nikolovlazar** [reply]
Not sure why that happens, but since the build command wasn't executed you don't see the Prose component in effect. I'll try to figure out why the command hangs and let you know. In the mean time if you find a solution let me know.

**Reply by nikolovlazar** [reply]
Can you try these commands one by one and let me know if it still hangs for you:
- `yarn clean`
- `yarn`
- `yarn build`
- `yarn storybook`

Run all of these at the root of the project.

**Reply by nikolovlazar** [reply]
I've opened a PR that includes the new `Prose` component. You check it out here: https://github.com/chakra-ui/chakra-ui/pull/5724 There's a Storybook deployment in which you can see how the styles look like:
https://chakra-ui-storybook-git-feature-prose-component-chakra-ui.vercel.app/?path=/story/components-typography-prose--basic

**Comment by nikolovlazar** [comment, marked as answer]
Hello @PrimozRome! I've some updates for you. I've published the `Prose` component as an individual npm package. Check it out and let me know what you think, or if you encounter some errors: https://github.com/nikolovlazar/chakra-ui-prose

**Reply by PrimozRome** [reply]
Excellent, I will give it a try in the next days on website I am working on and come back with a feedback. Thanks again!

**Comment by PrimozRome** [comment]
@nikolovlazar I have started to test this (I know a bit late). But `Prose` component does not render my HTML tags. The result I get is this:

<img width="507" alt="Screenshot 2022-06-07 at 11 20 06" src="https://user-images.githubusercontent.com/1180467/172344820-2b92695a-ac20-4437-bc90-3aac360c96ce.png">

HTML:

<img width="710" alt="Screenshot 2022-06-07 at 11 20 24" src="https://user-images.githubusercontent.com/1180467/172344890-004df4b7-f6f4-4fb9-858d-d709f52955b0.png">

Here is my code:

```
import {   Box,   Container, } from "@chakra-ui/react";
import { Prose } from "@nikolovlazar/chakra-ui-prose";

export default function ProductHeader({ name, title }) {
  return (
    <Box as="section" bg="blackAlpha.50" w="full" py="4">
      <Container maxW="container.2xl">
        <Prose>{title}</Prose>
      </Container>
    </Box>
  )
}
```

What am I doing wrong?

**Reply by nikolovlazar** [reply]
You should use the `dangerouslySetInnerHTML` property, like so:
```jsx
<Prose>
   <div dangerouslySetInnerHTML={{ _html: content }} />
</Prose>
```

**Reply by nikolovlazar** [reply]
The reason why that functionality got left outside of the package is because I want to leave room for the users to bring their own [content sanitization methods](https://github.com/nikolovlazar/chakra-ui-prose#content-sanitization).

**Reply by PrimozRome** [reply]
I understand. 

Should we update this in the [docs](https://chakra-ui.com/guides/recipes/prose)?

**Reply by nikolovlazar** [reply]
Oh we definitely should!

**Comment by PrimozRome** [comment]
@nikolovlazar I don't think it is 100% semantically correct that `<Prose>` component wraps the content inside the `<article>` tag?

Sometimes `<Prose>` content is a part of higher `<article>` and then you end up having `<article>` in `<article>` ...

Thoughts?

**Reply by nikolovlazar** [reply]
In cases like that you can use the `as` prop of the `Prose` component to change it. By default it's an `article` tag, but it can be anything you provide.

**Reply by PrimozRome** [reply]
Ah cool, thanks!

One more question. If I need multiple different styling sets for `<Prose>`, how would I approach that?

**Reply by nikolovlazar** [reply]
I haven't taken that into consideration, but I can look into this.

**Reply by PrimozRome** [reply]
@nikolovlazar that would be useful because Prose styles for blog posts can be very different than Prose styles for Product page description, for example...

Tailwind also allows direct default styles overrides on the Prose component https://tailwindcss.com/docs/typography-plugin#element-modifiers.

**Comment by PrimozRome** [comment]
I can't apply other style props to <Prose> like `w`, `px`, etc.?

**Reply by nikolovlazar** [reply]
Not directly on the `Prose` component, no.

**Reply by nikolovlazar** [reply]
@PrimozRome I've released version `1.2.0` of the Prose package that includes support for `BoxProps`, plus I've reduced its bundle size and made it tree shakeable: 
![CleanShot 2022-06-12 at 00 27 16](https://user-images.githubusercontent.com/5396211/173206857-62f7219a-e173-4eba-ab70-7608ede7fac9.png)

I'll think how to approach the multiple styling issue soon. Let me know how it goes 🙌

**Reply by PrimozRome** [reply]
Perfect, already on it!
